### PR TITLE
Clean tests & backwards compatibility tests.

### PR DESCRIPTION
**Pull request by stephenroller**
- Add backwards-compatibility tests for seq2seq, transformer ranker, transformer generator.
- Clean up testing framework, removing some redundant code (`_mock_train`)
- Make transformer tests main tests.
- Suppress the output from test_pytorch_data_teacher
- Decorator for marking tests retrying flaky tests.

### CODE REVIEW

**Comment by stephenroller** [comment]
Good for review.

File: parlai/core/testing_utils.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    Decorator for skipping a test if a GPU is available.
+
+    Useful for disabling hogwild tests.
+    """
+    return unittest.skipIf(GPU_AVAILABLE, reason)(testfn)
+
+
+def skipUnlessGPU(testfn, reason='Test requires a GPU'):
+    """Decorator for skipping a test if no GPU is available."""
+    return unittest.skipIf(not GPU_AVAILABLE, reason)(testfn)
```
Comment (klshuster) [inline parlai/core/testing_utils.py]: nit: considering you used the `skipUnless` syntax above, maybe good to be consistent?

File: parlai/core/testing_utils.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+    d = tempfile.mkdtemp()
+    yield d
+    shutil.rmtree(d)
+
+
+def train_model(opt):
+    """
+    Runs through a TrainLoop.
+
+    If model_file is not in opt, then this helper will create a temporary
```
Comment (klshuster) [inline parlai/core/testing_utils.py]: nit: create a temporary 'directory'

File: tests/integration/test_fairseq.py [hunk 3]
```diff
@@ -25,28 +19,12 @@
...
-    return stdout.getvalue(), valid, test
-
-
 class TestFairseq(unittest.TestCase):
     """Checks that fairseq can learn some very basic tasks."""
 
     @unittest.skipIf(SKIP_TESTS, "Fairseq not installed")
     def test_labelcands(self):
-        stdout, valid, test = _mock_train(
+        stdout, _, valid, test = testing_utils.train_model(dict(
```
Comment (klshuster) [inline tests/integration/test_fairseq.py]: why is there an additional `_` here - does `output.getvalue()` return a tuple?

File: tests/test_hogwild.py [hunk 4]
```diff
@@ -4,102 +4,68 @@
...
-        self.data.append(s)
+import parlai.core.testing_utils as testing_utils
 
-    def flush(self):
-        pass
-
-    def __str__(self):
-        return "".join(self.data)
+NUM_EXS = 100
+# ideally we wants something that's nice modulo with NUM_EXS and not
```
Comment (klshuster) [inline tests/test_hogwild.py]: nit: appears to be unfinished comment

File: tests/test_memnn.py [hunk 5]
```diff
@@ -5,42 +5,20 @@
...
-
 class TestMemnn(unittest.TestCase):
     """Checks that seq2seq can learn some very basic tasks."""
 
     def test_labelcands_nomemnn(self):
         """This test uses a single-turn task, so doesn't test memories."""
 
-        stdout, valid, test = _mock_train(
+        stdout, valid, test = testing_utils.train_model(dict(
             task='integration_tests:CandidateTeacher',
```
Comment (klshuster) [inline tests/test_memnn.py]: nit: i notice in later files, you substitute `CandidateTeacher` for `candidate`, maybe do so here to keep consistent style?

File: tests/integration/test_fairseq.py [hunk 6]
```diff
@@ -68,7 +46,7 @@ def test_labelcands(self):
 
     @unittest.skipIf(SKIP_TESTS, "Fairseq not installed")
     def test_generation(self):
-        stdout, valid, test = _mock_train(
+        stdout, _, valid, test = testing_utils.train_model(dict(
```
Comment (klshuster) [inline tests/integration/test_fairseq.py]: ditto

**Pull request merged**